### ISSUE TITLE
Questions about tfms and unexpected decoder behaviour with `TSStandardize`

### ISSUE THREAD

**Issue by diegoquintanav** [timeseriesAI/tsai#195]
Considering the bit in the following [gist](https://colab.research.google.com/gist/diegoquintanav/4c33801c9167cde4feab3beaaa3f42ce/scratchpad.ipynb) 

```python
tfms  = [None, [TSForecasting()]] # breaks other methods
batch_tfms = TSStandardize(by_sample=False, by_var=True, verbose=True)
dls_inc = get_ts_dls(X, y, splits=splits, tfms=tfms, batch_tfms=batch_tfms, bs=128, device="cpu")
```

After I instantiate and train a learner, I can use `learner.get_X_preds()`, which returns preds and decoded preds, which are the same. I am expecting that the decoded values are the preds, but scaled back i.e. the inverse standardization is applied to them.

If you check the notebook in the end, this is not happening. I understand that from fastai's API, transformations implement a decode method such as [Normalize](https://docs.fast.ai/data.transforms.html#Normalize). I have many questions :sweat_smile:  

- How can I revert the standardization on the preds? in other words, that `decoded_preds` is an array of preds with the inverse scaling applied to it
- What are the means showing up in `dls_inc.after_batch.mean.data`? What is the proper way to scale back a tensor using `dls` or `learner`?
- Related to the previous question, what are `after_batch`, `before_batch`, `after_iter`, `before_iter` in the transforms? 

Thanks again for your help!

**Comment by oguiza** [comment]
Hi @diegoquintanav,

I'll try to clarifiy at least some of your questions.

When you train a model in tsai, the dataloader needs to scale the X values (that is that batch tfms == after_batch do). In this case you are using TSStandardize. It only modifies the X values. 

In a forecasting task, the y (target) is not really transform. The only transformation is to make sure data is formatted as floats.

The model is thus trained to create the target (ys)  that you pass using the standardized inputs (standardized Xs). The output the model produces during training is getting closer and closer to the target. But again, the target is never changed. 

When training finishes and you use learn.get_X_preds, you pass the Xs (unscaled), and optionally the (ys). The function then scales the Xs, pass them through the model, and generates the output y hat. 

In your gist y_hat are very far from the target because you are using random data, so the model cannot learn anything. But if you pass real data, the model learns pretty well, and generate good predictions. 

Since no tfms are applied to the ys, in forecasting and regression tasks, preds and decoded preds are the same values. 

As to your questions on ```after_batch```,```before_batch```, ```after_iter```,```before_iter``` the only ones you'll ever use are: 

- tfms = after_item (these are transforms applied to each individual item - usually the ys-, like TSClassification, that transforms each label in a digit) and  
- batch_tfms = after_batch that are the transforms applied to each individual batch once it's been created.

I hope this is a bit clearer now 🙂

**Comment by diegoquintanav** [comment]
Hi @oguiza 

As always thank you for your thorough response! I did play with the learning rate and the epochs and also with a different dataset and got better predictions, in the scale of the input. In any case, how would you scale back any tensor using a ´dls´ or `learner` instance?. 
thanks again!

**Comment by oguiza** [comment]
I'd do it this way: 


```
X, y, splits = get_UCR_data('LSST', return_split=False)
dls = get_ts_dls(X, y, splits=splits, tfms=[None, TSClassification()], bs=[64, 128], batch_tfms=TSStandardize())
xb,yb=dls.one_batch()
idxs = dls.input_idxs 
X[idxs].mean(), X[idxs].std(), xb.mean(), xb.std()
```

```
dec_xb, dec_yb = dls.decode((xb, yb))
dec_xb.mean(), dec_xb.std()
```

**Comment by diegoquintanav** [comment]
Wow, I was complicating myself. Thanks again.
